src/meta/src/hummock/manager/mod.rs

```diff
@@ -2209,20 +2066,23 @@ where
                                         hummock_manager.compactor_manager.clone();
 
                                     // TODO: add metrics to track expired tasks
-                                    for (context_id, mut task) in
-                                        compactor_manager.get_expired_tasks()
+                                    const INTERVAL_SEC: u64 = 30;
+                                    // The cancel task has two paths
+                                    // 1. compactor heartbeat cancels the expired task based on task
+                                    // progress (meta + compactor)
+                                    // 2. meta periodically scans the task and performs a cancel on
+                                    // the meta side for tasks that are not updated by heartbeat
```

Comment:
Btw, if it is case 1 (not receiving a heartbeat), I think we can drop the task directly since the current protocol guaranteed that all alive tasks will be included in the task progress list on each heartbeat.